app/grandchallenge/notifications/migrations/0008_migrate_forum_follows_and_notifications.py

```diff
@@ -0,0 +1,111 @@
+from django.db import migrations
+
+from grandchallenge.notifications.models import NotificationTypeChoices
+
+
+def migrate_forum_and_topic_follows_and_notifications(apps, schema_editor):
+    Notification = apps.get_model(  # noqa: N806
+        "notifications", "Notification"
+    )
+    Follow = apps.get_model("actstream", "Follow")  # noqa: N806
+    Forum = apps.get_model("discussion_forums", "Forum")  # noqa: N806
+    ForumTopic = apps.get_model(  # noqa: N806
+        "discussion_forums", "ForumTopic"
+    )
+    MachinaForum = apps.get_model("forum", "Forum")  # noqa: N806
+    MachinaTopic = apps.get_model("forum_conversation", "Topic")  # noqa: N806
+    ContentType = apps.get_model("contenttypes", "ContentType")  # noqa: N806
+
+    new_topic_ct = ContentType.objects.get_for_model(ForumTopic)
+    new_forum_ct = ContentType.objects.get_for_model(Forum)
+
+    def get_matching_forum(*, old_forum_id):
+        old_forum = MachinaForum.objects.get(pk=old_forum_id)
+        return old_forum.challenge.discussion_forum
+
+    def get_matching_topic(*, old_topic_id):
+        old_topic = MachinaTopic.objects.get(pk=old_topic_id)
+        new_forum = get_matching_forum(old_forum_id=old_topic.forum.pk)
+        return ForumTopic.objects.get(
+            forum=new_forum,
+            creator=old_topic.poster,
+            subject=old_topic.subject,
+        )
+
+    notifications_to_update = []
+    for notification in Notification.objects.filter(
+        type__in=[
+            NotificationTypeChoices.FORUM_POST,
+            NotificationTypeChoices.FORUM_POST_REPLY,
+        ]
+    ):
+        if notification.type == NotificationTypeChoices.FORUM_POST:
+            new_target = get_matching_forum(
+                old_forum_id=notification.target_object_id
+            )
+            new_action_object = get_matching_topic(
+                old_topic_id=notification.action_object_object_id
+            )
```

Comment:
How do we know if a notification is migrated? Does this run atomically?

Reply:
No longer relevant if we don't migrate them anymore, but just for my understanding: my thinking was that we run this migration once and so each object gets updated once, and so we don't need to know if it's been migrated or not. 

If a notification has been migrated though, `get_matching_forum` would fail, so we could catch it with a try ... except block and skip notifications for which we can't retrieve a challenge forum (that would also cover the General Forum notifications).